### PR TITLE
Fix: Add ECDSA certificate support for internal TLS (#20082)

### LINKED ISSUE #20082: ECDSA certificates for internal tls aren't supported
**Expected behavior and actual behavior:**

I was using cert-manager with this:

```
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: internal-tls-ca
spec:
  isCA: true
  commonName: internal-tls-ca
  secretName: internal-tls-ca
  privateKey:
    algorithm: ECDSA
    size: 256
  issuerRef:
    name: self-signed
    kind: ClusterIssuer
    group: cert-manager.io
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: internal-tls-issuer
spec:
  ca:
    secretName: internal-tls-ca
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: harbor-core-internal-tls
spec:
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: internal-tls-issuer
  commonName: harbor-core
  dnsNames:
    - localhost
    - harbor-core
  ipAddresses:
    - 127.0.0.1
  usages:
    - server auth
    - client auth
  secretName: harbor-core-internal-tls
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: harbor-jobservice-internal-tls
spec:
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: internal-tls-issuer
  commonName: harbor-jobservice
  dnsNames:
    - harbor-jobservice
  usages:
    - server auth
    - client auth
  secretName: harbor-jobservice-internal-tls
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: harbor-registry-internal-tls
spec:
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: internal-tls-issuer
  commonName: harbor-registry
  dnsNames:
    - harbor-registry
  usages:
    - server auth
    - client auth
  secretName: harbor-registry-internal-tls
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: harbor-portal-internal-tls
spec:
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: internal-tls-issuer
  commonName: harbor-portal
  dnsNames:
    - harbor-portal
  usages:
    - server auth
    - client auth
  secretName: harbor-portal-internal-tls
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: harbor-trivy-internal-tls
spec:
  issuerRef:
    group: cert-manager.io
    kind: Issuer
    name: internal-tls-issuer
  commonName: harbor-trivy
  dnsNames:
    - harbor-trivy
  usages:
    - server auth
    - client auth
  secretName: harbor-trivy-internal-tls
```

It generally works, components can communicate with each other, Portal is operational etc etc. But pulling public images or logging in with `docker login` do not work (note that in the currently released version i.e. `v2.10.0` there is a separate issue https://github.com/goharbor/harbor/issues/20080 which makes symptoms entirely different, core returns 200 response with an empty token; this is what I get with my patch from https://github.com/goharbor/harbor/pull/20081):

```
> nerdctl pull harbor.example.com/hub.docker.com/jenkins/jenkins:lts
harbor.example.com/hub.docker.com/jenkins/jenkins:lts: resolving      |--------------------------------------|
elapsed: 2.4 s                                                            total:   0.0 B (0.0 B/s)
INFO[0002] trying next host                              error="failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://harbor.example.com/service/token?scope=repository%3Ahub.docker.com%2Fjenkins%2Fjenkins%3Apull&service=harbor-registry: 500 Internal Server Error" host=harbor.example.com
FATA[0002] failed to resolve reference "harbor.example.com/hub.docker.com/jenkins/jenkins:lts": failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://harbor.example.com/service/token?scope=repository%3Ahub.docker.com%2Fjenkins%2Fjenkins%3Apull&service=harbor-registry: 500 Internal Server Error
Error: exit status 1
```

Core logs ():

```
2024-03-03T22:45:39Z [ERROR] [/core/service/token/token.go:50]: Unexpected error when creating the token, error: x509: failed to parse private key (use ParseECPrivateKey instead for this key format)
```

**Steps to reproduce the problem:**

Try to pull public image or log in to Harbor using ECDSA certificate as internal tls CA.

**Versions:**
Please specify the versions of following systems.

- harbor version: [2.10.0]

**Additional context:**

Slack thread: https://cloud-native.slack.com/archives/CC1E09J6S/p1709497679881159

### PR DESCRIPTION
**Pull request by rossigee**
## Description

Adds ECDSA private key support to Harbor's internal token signing flow, addressing issue #20082.

### Problem
Harbor's token service failed to parse ECDSA private keys with the error: `x509: failed to parse private key`. This prevented users from using ECDSA certificates with cert-manager for Harbor's internal TLS.

### Solution
Enhanced the token package to provide full ECDSA support equivalent to existing RSA functionality.

### Core Functionality
- **Automatic Algorithm Detection**: NewOptions detects key type and selects JWT method:
  - P-256 → ES256, P-384 → ES384, P-521 → ES512
- **Multi-format PEM Support**: SEC1 (`EC PRIVATE KEY`) and PKCS#8 (`PRIVATE KEY`)
- **Key Validation**: Public/private key consistency checks for ECDSA matching RSA behavior

### Technical Details
- **NewOptions**: Enhanced to detect and parse ECDSA keys, proper PEM decoding across multiple blocks
- **GetKey**: Added ECDSA case with full key validation and public/private matching
- **Parse**: Updated with key-type-derived algorithm allowlist (RSA: RS256/384/512/PS256/384/512, ECDSA: ES256/384/512)
- **Linting**: Uses // nolint:revive to suppress revive package name checks on token packages

### Testing & Quality
- **Comprehensive Tests**: Automated ECDSA test coverage with all edge cases
- **Format Coverage**: SEC1, PKCS8 for P-256, P-384, P-521 curves
- **Key Validation**: Matching/mismatched public-private key rejection
- **Error Handling**: Missing keys, malformed keys, unsupported formats
- **Backward Compatibility**: Existing RSA configurations work unchanged

## Files Changed
- `src/pkg/token/options.go` - ECDSA key detection, parsing, validation
- `src/pkg/token/option_test.go` - Comprehensive ECDSA test coverage
- `src/pkg/token/token.go` - Key-type algorithm allowlist
- `src/core/service/token/token_test.go` - Enhanced error handling in helper functions
- `src/.golangci.yaml` - Linting directive for token package

## Backward Compatibility
✅ **No breaking changes** - Existing RSA configurations and code continue to work unchanged

## Testing
```bash
cd src && go test ./pkg/token/... -v    # All token tests pass
cd src && go test ./core/service/token/... -v  # Integration tests pass with DB setup
```

## Sign-off
Signed-off-by: Ross Golder <ross@golder.org>

## Related Issue
Fixes #20082 - ECDSA certificates for internal tls aren't supported

### CODE REVIEW

**Comment by Vad1mo** [comment]
@rossigee DCO is missing

**Comment by rossigee** [comment]
Rebased and addressed review feedback:

- All commits now have Signed-off-by (DCO)
- Fixed  cleanup order in  (before )
- Added coverage tests for previously untested error paths in :
  - RSA/ECDSA private key parse errors in GetKey
  - ECDSA public key parse error in GetKey  
  - Unsupported sign method in GetKey
  - Unsupported ECDSA curve (P-224) in NewOptions
  - Unsupported key type (Ed25519) in NewOptions
- Coverage on options.go is now 100% (package: 89.4%)
- Regenerated API files from swagger spec
- Fixed golangci.yaml indentation

**Comment by rossigee** [comment]
Rebased and addressed review feedback:

- All commits now have Signed-off-by (DCO)
- os.Remove cleanup now happens before os.Exit in TestMain
- Added coverage tests for previously untested error paths in options.go
- options.go coverage is now 100% (package: 89.4%)
- Regenerated API files from swagger spec
- Fixed golangci.yaml indentation

**Comment by rossigee** [comment]
What do I need to do for this 'release note label check' failure?

**Comment by rossigee** [comment]
## PR Status Update

✅ **DCO Issue - RESOLVED**: All commits now include proper `Signed-off-by` sign-offs per Developer Certificate of Origin requirements.

⚠️ **Release Note Label Needed**: This PR requires a `release-note/*` label to pass the release note label check. 

Since this is a bug fix for ECDSA certificate support for internal TLS, I recommend adding the **`release-note/update`** label.

@goharbor/maintainers - Could a maintainer please add the appropriate release-note label? Thank you!

**Comment by rossigee** [comment]
## ✅ All Copilot Review Comments Resolved

All feedback from Copilot reviews has been addressed:

### Code Quality Fixes
- ✅ **Error formatting**: Using `%v` for non-string values in error messages
- ✅ **Error wrapping**: Proper use of `%w` for error chain preservation
- ✅ **Grammatical fixes**: Corrected error messages to "do not match"

### Security & Robustness
- ✅ **PEM block iteration**: Handles multi-block PEM files (e.g., EC PARAMETERS + EC PRIVATE KEY)
- ✅ **Whitespace handling**: Trims whitespace when scanning PEM blocks to avoid false negatives
- ✅ **Curve comparison**: Uses `.Curve.Params().Name` for robust ECDSA curve matching
- ✅ **Key equality**: Uses `.Equal()` method instead of interface identity comparison

### Test Quality
- ✅ **File close errors**: All `Close()` calls are properly error-checked
- ✅ **Environment isolation**: TestMain now restores `TOKEN_PRIVATE_KEY_PATH` after tests
- ✅ **Comprehensive coverage**: 
  - ECDSA support (P-256, P-384, P-521 curves)
  - Both SEC1 and PKCS8 key formats
  - Error path coverage for key parsing
  - Integration tests through token creation/parsing
  - Multi-block PEM file support
  - Unsupported key type detection (P-224, Ed25519)

### Configuration & Documentation
- ✅ **YAML validation**: golangci.yaml properly formatted
- ✅ **Code documentation**: NewOptions function documents deprecated parameter
- ✅ **Error context**: Error messages include file paths for better debugging

### Code Changes
The final commit addresses the last remaining issue:
- **Commit 9994d36d1**: Restore environment variable in TestMain cleanup

All tests pass with race condition detection enabled.

File: src/pkg/token/options.go [hunk 1]
```diff
@@ -15,9 +15,14 @@
 package token // nolint:revive
```
[the remark this answers is not in the document]
Comment (rossigee) [reply src/pkg/token/options.go:15]: ✅ Resolved - Updated PR description to accurately reflect the implementation using nolint:revive directives instead of package renaming.

File: src/pkg/token/token.go [hunk 2]
```diff
@@ -65,13 +66,40 @@ func Parse(opt *Options, rawToken string, claims jwt.Claims) (*Token, error) {
...
+	switch key.(type) {
+	case *rsa.PrivateKey, *rsa.PublicKey:
+		validMethods = []string{
+			jwt.SigningMethodRS256.Alg(),
+			jwt.SigningMethodRS384.Alg(),
+			jwt.SigningMethodRS512.Alg(),
+			jwt.SigningMethodPS256.Alg(),
+			jwt.SigningMethodPS384.Alg(),
+			jwt.SigningMethodPS512.Alg(),
+		}
```
[the remark this answers is not in the document]
Comment (rossigee) [reply src/pkg/token/token.go:82]: ✅ Resolved - Updated PR description to reflect that RS* and PS* (PSS) variants are supported for RSA keys alongside ES* for ECDSA. Also updated the token.go comment to clarify PS256/384/512 are accepted.

File: src/pkg/token/token.go [hunk 3]
```diff
@@ -65,13 +66,37 @@ func Parse(opt *Options, rawToken string, claims jwt.Claims) (*Token, error) {
...
+	// and RSA (RS256/384/512) keys are each accepted regardless of which specific
+	// variant was used when the token was issued.
+	var validMethods []string
+	switch key.(type) {
+	case *rsa.PrivateKey, *rsa.PublicKey:
+		validMethods = []string{
+			jwt.SigningMethodRS256.Alg(),
+			jwt.SigningMethodRS384.Alg(),
+			jwt.SigningMethodRS512.Alg(),
+		}
```
[the remark this answers is not in the document]
Comment (rossigee) [reply src/pkg/token/token.go:82]: ✅ Resolved - Updated the Parse() method's validMethods allowlist to include PS256/384/512 for RSA keys, making it consistent with GetKey() which already supports jwt.SigningMethodRSAPSS.

File: src/pkg/token/token.go [hunk 4]
```diff
@@ -65,13 +66,37 @@ func Parse(opt *Options, rawToken string, claims jwt.Claims) (*Token, error) {
...
+	case *ecdsa.PrivateKey, *ecdsa.PublicKey:
+		validMethods = []string{
+			jwt.SigningMethodES256.Alg(),
+			jwt.SigningMethodES384.Alg(),
+			jwt.SigningMethodES512.Alg(),
+		}
```
[the remark this answers is not in the document]
Comment (rossigee) [reply src/pkg/token/token.go:88]: ✅ Resolved - The algorithm allowlist expansion is intentional: it allows tokens signed with any variant (RS/PS/ES) of a given key type. This is the correct behavior for token parsing. Added clarifying comment to document this allowlist strategy.

**Comment by rossigee** [comment]
## ✅ Copilot Review - All Issues Addressed

### Resolved Comments
**Direct Replies Posted:**
- [#3504310618](https://github.com/goharbor/harbor/pull/22861#discussion_r3504310618) - PR description updated to reflect nolint:revive implementation
- [#3504361752](https://github.com/goharbor/harbor/pull/22861#discussion_r3504361752) - PS256/384/512 support clarified in PR description
- [#3504361947](https://github.com/goharbor/harbor/pull/22861#discussion_r3504361947) - Parse() validMethods includes PS* variants
- [#3504362067](https://github.com/goharbor/harbor/pull/22861#discussion_r3504362067) - Algorithm allowlist intentional and documented

### Code Fixes Applied
✅ Multi-block PEM support working (scans blocks, stores matched block only)  
✅ Error handling with %w wrapping and file paths in messages  
✅ Whitespace trimming implemented for edge cases  
✅ Deprecated parameter accepted for backward compatibility  
✅ TestMain env cleanup properly restores prior values  
✅ GetKey() assertions added to PKCS8 test  
✅ Enhanced documentation on NewOptions deprecation  
✅ PS256/384/512 support verified in TestParseWithPS256  

### Test Coverage
- ✅ 18 token package tests all passing
- ✅ Multi-format support: SEC1 & PKCS8
- ✅ Multi-curve support: P-256, P-384, P-521
- ✅ ECDSA integration tests (MakeToken, Parse)
- ✅ RSAPSS (PS256) support verified
- ✅ All CI checks COMPLETED

Ready for merge! 🚀

File: src/pkg/token/options.go [hunk 5]
```diff
@@ -62,41 +65,116 @@ func (o *Options) GetKey() (any, error) {
...
-// NewOptions create Options based on input parms
-func NewOptions(sm, iss, keyPath string) (*Options, error) {
-	signMethod := jwt.GetSigningMethod(sm)
-	if signMethod == nil {
-		return nil, fmt.Errorf("unknown sign method: %s", sm)
+// NewOptions creates Options based on the input parameters.
+// The first parameter is deprecated and ignored; the signing method is
+// automatically determined from the key type.
+func NewOptions(_, iss, keyPath string) (*Options, error) {
+	pkBytes, err := os.ReadFile(keyPath)
```
[the remark this answers is not in the document]
Comment (rossigee) [reply src/pkg/token/options.go:120]: ✅ Resolved - Added deprecation logging when non-empty method parameter is passed. This provides visibility to callers that the parameter is ignored while maintaining backward compatibility. Callers passing explicit methods (RS512, PS256) will see a debug log message indicating the method is auto-detected from key type instead.
